### DISCUSSION TITLE
keep columns centered

### DISCUSSION THREAD

**Discussion by gregzuro** [YaLTeR/niri#381] [Ideas]
If I center a column with `center-column`, then it might be nice to keep it in the center if the width is changed with `set-column-width`.

**Comment by morenathan** [comment]
It might actually be nice to have the concept of, "pin to center". I myself rarely use the center feature but when I do it's purposely meant for a centered window.

**Comment by YaLTeR** [comment]
This is what already happens with:
```
layout {
    center-focused-column "always"
}
```
along with a number of other always-center-specific things (like horizontal gesture snapping).

I don't want to add these behaviors otherwise because then you get into a ton of nasty edge cases with when you consider a column in the "centering mode".

**Reply by gregzuro** [reply]
If #377 is fixed, that may be what I'm after.

**Reply by morenathan** [reply]
@YaLTeR I've not tried binding that to a key honestly. Will it toggle?

Seems like right now the "center-column" call was meant to tried a layout reshuffle instead of just temporarily centering the window until you do something.

Maybe "center-column" should work at the layout level instead of the window level and change the layout to use "center-focused-column" until toggled back off?

@gregzuro sorry if I've shoved this off the rails. I'm also not sure how that solves the problem you brought up here?

**Comment by gregzuro** [comment]
> Maybe "center-column" should work at the layout level instead of the window level and change the layout to use "center-focused-column" until toggled back off?

You may be on to something with this...

**Reply by morenathan** [reply]
The problem is, center-focused-column doesn't distinguish between screens as far as I can tell. I don't really want all my displays suddenly centered, or not centered. I want only the screen with the centered window to receive this treatment.

ATM screen-level layouts are not implemented as far as I can tell.

**Reply by YaLTeR** [reply]
I frequently use center-column without the intention of going full centering mode.

Btw you can already make a bind like this: spawn a script that changes the niri config with `sed` to comment/uncomment the center always setting.

**Reply by morenathan** [reply]
The only reason I find that a disheartening answer is; 

1, it doesn't prevent other monitors from center as you move between them, in fact, they WILL center the active window; 

2, Because, you then have to be toggling this option on and off to avoid #1.

Not that it's for me to decide the features of this Compositor. Although, even in the short run it's not a truly good solution. I have tested this solution, at least manually, and it 'works' until you switch monitors while it's on.

Or did I miss something?

**Reply by YaLTeR** [reply]
> it doesn't prevent other monitors from center as you move between them, in fact, they WILL center the active window;

Hm, right. I guess this option could be per-monitor (but we need to decide on the best place to put per-monitor options).

**Reply by morenathan** [reply]
My thought was that each "display"  in the config could have its own "layout". This would be instead of a single global layout.

**Comment by morenathan** [comment]
To indulge myself a little, maybe I could demonstrate my response from yesterday? I would see the configuration changing to look something more like this;

```
// <identifier> could be anything from the /sys/drm name
// to anything supported in the future.
output "<identifier>" {
	scale 1.0
	transform "90"
	mode "<width>x<height>"
	position x=0 y=0
	
	// this is a screen-specfic layout override
	layout {
		center-focused-column "always"
		gaps 0
		struts {
			left 0
			right 0
			top 0
			bottom 0
		}
		default-column-width { proportion 1.00; }
		focus-right { off; }
		border { off; }
	}
}

output "<identifier>" {
	scale 1.0
	transform "normal"
	mode "<width>x<height>"
	position x=<last_width> y=0
	// ### no layout override ###
}

layout {
	center-focused-column "always"
	gaps 5
	struts {
		left 5
		right 5
		top 5
		bottom 5
	}
	default-column-width { proportion 0.50; }
	preset-column-widths {
		proportion 0.75
		proportion 1.00
		proportion 0.25
	}
	focus-right { off; }
	border { 
		width 5
		active-color 255 0 0 255
		inactive-color 0 0 0 0
	}
}
```

So that's how I would imagine this working. My 20₡ worth. Hope this doesn't put you off-guard at all. EDIT: made some minor changes.

**Reply by YaLTeR** [reply]
Perhaps, yeah. There was some other request to customize the default width for specific outputs, too.

Would need refactoring to deal with merging together `layout` configs, and some thinking on how to allow the `on` flag inside output without duplicating all types involved.

**Reply by morenathan** [reply]
At least these settings would only apply at startup. Although it would be nice if you could have separate layout sections with name that you could bind to a key.

**Reply by morenathan** [reply]
I'm glad you think this is something worthwhile. For a single-screen computer I don't know how useful this would be, but for a multi-monitor setup it's definitely a plus.

IIRC it's also a feature in the paperwm plugin for gnome. Which I've really enjoyed using in the past. My standard for this type of window manager as always wmii, which is just simply an amazing wm. It's stuck in X though and I wouldn't know where to begin to start porting it to wayland.

**Reply by YaLTeR** [reply]
Made an issue: https://github.com/YaLTeR/niri/issues/401
